service/controller/release/resource/app/desired.go

```diff
@@ -0,0 +1,121 @@
+package app
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	applicationv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/application/v1alpha1"
+	releasev1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/release/v1alpha1"
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/release-operator/pkg/project"
+	"github.com/giantswarm/release-operator/service/controller/key"
+)
+
+func (r *resourceStateGetter) GetDesiredState(ctx context.Context, obj interface{}) ([]*applicationv1alpha1.App, error) {
+	cr, err := key.ToReleaseCR(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var desiredComponents []releasev1alpha1.ReleaseSpecComponent
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding desired components for release %#q", cr.Name))
+
+		desiredComponents, err = r.getDesiredComponents(ctx, cr)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d desired components for release %#q", len(desiredComponents), cr.Name))
+	}
+
+	var desiredAppCRs []*applicationv1alpha1.App
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("computing desired app CRs for release %#q components", cr.Name))
+
+		for _, c := range desiredComponents {
+			appCR := newAppCR(appCRName(c), c.Name, c.Version, r.appCatalog)
+			desiredAppCRs = append(desiredAppCRs, appCR)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("computed %d desired app CRs for release %#q components", len(desiredAppCRs), cr.Name))
+	}
+
+	return desiredAppCRs, nil
+}
+
+func (r *resourceStateGetter) getDesiredComponents(ctx context.Context, cr *releasev1alpha1.Release) ([]releasev1alpha1.ReleaseSpecComponent, error) {
+	// If this is a non-EOL release and it isn't deletion event all
+	// components of that release are desired.
+	cr.GetDeletionTimestamp()
+	if cr.Status.Cycle.Phase != releasev1alpha1.CyclePhaseEOL && !key.IsDeleted(cr) {
+		return cr.Spec.Components, nil
+	}
+
+	var activeReleases []releasev1alpha1.Release
+	{
+		opts := metav1.ListOptions{
+			LabelSelector: key.LabelReleaseCyclePhase + "!=" + releasev1alpha1.CyclePhaseEOL.String(),
+		}
+
+		result, err := r.g8sClient.ReleaseV1alpha1().Releases("").List(opts)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		activeReleases = result.Items
+	}
+
+	var desiredComponents []releasev1alpha1.ReleaseSpecComponent
+
+	// If this an EOL release only components that are also part of other
+	// non-EOL (active) releases are desired.
```

Comment:
Why? What is usually obviously expressed by the code, the why is not. So let's please write down the cause.

Reply:
Would something like that work?

> If this an EOL release components that are also part of other non-EOL (active) releases are still desired. Otherwise active releases consisting of those shared components would stop working till next reconciliation loop.

Reply:
That makes sense to me up to the point of the next reconciliation loop. That opens more questions. But yeah, let'S extend the comment a bit and go.

Reply:
Then maybe:

> If all App CRs for components of this EOL release were blindly deleted there is a chance some active (non-EOL) release sharing some of those components would stop working (that would be eventually fixed when the broken release is reconciled again and App CRs for missing components are recreated). To avoid such situations this release components that are also part of other active releases are still desired.

Reply:
`situations this release` -> `situations these release`

Reply:
`To avoid such situations this release components that are also part of other active releases are still desired`

`this release components` - meaning the components of the (this) Release CR that is already reconciled. I meant `this release` not `these ... components`. Maybe `this release's components` is more accurate.

Reply:
Then it is `this release's components` IMO.

Reply:
Done.